AppDB/appscale/datastore/fdb/transactions.py

```diff
@@ -0,0 +1,305 @@
+"""
+This module stores and retrieves datastore transaction metadata. The
+TransactionManager is the main interface that clients can use to interact with
+the transaction layer. See its documentation for implementation details.
+"""
+from __future__ import division
+
+import logging
+import math
+import random
+import sys
+from collections import defaultdict
+
+import six
+import six.moves as sm
+from tornado import gen
+
+from appscale.common.unpackaged import APPSCALE_PYTHON_APPSERVER
+from appscale.datastore.dbconstants import BadRequest, InternalError
+from appscale.datastore.fdb.codecs import (
+  decode_str, encode_versionstamp_index, Int64, Path, Text, TransactionID)
+from appscale.datastore.fdb.utils import (
+  DS_ROOT, fdb, MAX_ENTITY_SIZE, ResultIterator, VERSIONSTAMP_SIZE)
+
+sys.path.append(APPSCALE_PYTHON_APPSERVER)
+from google.appengine.datastore import entity_pb
+
+logger = logging.getLogger(__name__)
+
+
+class TransactionMetadata(object):
+  """
+  A TransactionMetadata directory handles the encoding and decoding details for
+  transaction metadata for a specific project.
+
+  The directory path looks like (<project-dir>, 'transactions').
+  Within this directory, keys are encoded as
+  <scatter-byte> + <txid> + <rpc-type (optional)> + <rpc-details (optional)>.
+
+  The <scatter-byte> is a single byte derived from the txid. Its purpose is to
+  spread writes more evenly across the cluster and minimize hotspots. This is
+  especially important for this index because each write is given a new, larger
+  <txid> value than the last.
+
+  The <txid> is an 8-byte integer that serves as a handle for the client to
+  identify a transaction. It also serves as a read versionstamp for FDB
+  transactions used within the datastore transaction.
+
+  The initial creation of the datastore transaction does not specify any RPC
+  details. The purpose of that KV is to verify that the datastore transaction
```

Comment:
Key-value, I see, but imho it would be better to keep the space for guesses smaller where possible, as there's already enough complexity to keep in mind.

Reply:
Oh.. I just noticed that I already was commenting on "KV". Sorry for repeating myself.

Reply:
I've pushed an update that changes a few more instances of "KV" in the docstrings to "Key-Value".